File: collatz.cs
using System;

struct numbers
{
    public long nums;
    public long steps;
}

class Collatz
{

    static void Main(string [] args)
    {
        int size = 10; 
        numbers [] c = new numbers[size];
        long max;
        Console.WriteLine("Please enter in a positive number");
        max = Convert.ToInt32(Console.ReadLine());

        // fill array
        for(long i = 0; i < size;i++)
        {
            c[i].nums = 0;
            c[i].steps = 0;
        }
        while(max != 0)
        {
            long stepnum = nSteps(max);
            bool dup = false;
            long smallest = 0;
            long small = c[0].steps;
            long dupIndex = 0;
            // find the smallest value 
            for(long i = 0; i < size; i++)
            {
                if(c[i].steps < small)
                {
                    small = c[i].steps;
                    smallest = i;
                }
                // if what is in the array is equal to the list say a duplicate is found  
                if(c[i].steps == stepnum)
                {
                    dup = true;
                    dupIndex = i;
                }
            }
            long index = smallest;
            // a duplicate is not found, then add numbers to the arrays 
            if(stepnum > c[index].steps && dup==false)
            {
                c[index].steps = stepnum;
                c[index].nums = max;
            }
            // if a duplicate is found, add only the number 
            if(max < c[dupIndex].steps && dup==true)
            {
                c[dupIndex].nums = max;
            }
            max--;
        }
        // sort based on seq Length 
        Console.WriteLine("Sorted by sequence length");
        Array.Sort(c,delegate(numbers a, numbers b) {return a.steps.CompareTo(b.steps);});
        for(long i = size-1; i>=0; i--)
        {
            Console.WriteLine(c[i].nums + "    " + c[i].steps);
        }
        // Sort based on integer Length 
        Console.WriteLine("Sorted by integer length");
        Array.Sort(c,delegate(numbers a, numbers b) {return b.nums.CompareTo(a.nums);});
        for(long i = 0; i <size ; i++)
        {
            Console.WriteLine(c[i].nums + "    " + c[i].steps);
        }
    }
     // calculate collatz conjecture 
    static long nSteps(long n)
    {
        long count = 0;
        if(n == 1) 
        {
            return count; 
        }
        else if ( (n & 1)==0 )
        {
            count = nSteps(n>>1) + 1; 
        }
        else
        {
            count = nSteps((n<<1)+n+1)+1;
        }
        return count; 
    }
}

File: collatz.java
import java.util.*;
public class collatz
{
    public static void main(String[] args)
    {
        long max;
        System.out.println("Please enter a positive number");
        Scanner input = new Scanner(System.in);
        max = input.nextLong();
        int size = 10; 
        long [] num = new long[size];
        long [] step = new long[size]; 
        long temp; 
        // inialize array 
        for(int i = 0; i < size; i++)
        {
            num[i] = 0;
            step[i] = 0;
        }
        while(max != 0)
        {
            long stepnum = numSteps(max);
            boolean dup = false;
            long smallest = 0;
            long small = step[0];
            int dupIndex = 0;
           for(int i = 0; i < size; i++)
            {
                // find smallest number 
                if(step[i] < small)
                {
                   small = step[i];
                   smallest = i;
                }
                // if a number is seen more than once, a duplicate is there 
                if(step[i] == stepnum)
                {
                    dup = true;
                    dupIndex = i;
                }
            }
            int index = (int)smallest;
            // if a duplicate is not there, include both numbers 
            if(stepnum > step[index] && dup == false)
            {
                step[index] = stepnum;
                num[index] = max;
            }
            // if a duplicate is found, do not add it 
            if(max < step[dupIndex] && dup == true)
            {
                num[dupIndex] = max;
            }
            max--;
        }
           // bubble sort 
           for(int j = 0; j < (size -1); j++)
            for(int k = 0; k < (size - j -1);k++)
                 if(step[k] > step[k+1])
                 {
                    temp = num[k];
                    num[k] = num[k+1];
                    num[k+1] = temp;

                    temp = step[k];
                    step[k] = step[k+1];
                    step[k+1] = temp;
                 }

          System.out.println("Sorted based on sequence length");
        
        for(int i = size-1; i >=0 ; i--)
        {
            System.out.println(num[i] + "   " + step[i]);
        }
        // bubble sort 
        System.out.println("Sorted based on integer length");
          for(int j = 0; j < (size-1); j++)

            for(int k = 0; k < (size-j-1);k++)

                if(num[k] > num[k+1])
                {
                   temp = step[k];
                   step[k] = step[k+1];
                   step[k+1] = temp;

                   temp = num[k];
                   num[k] = num[k+1];
                   num[k+1] = temp;
                }
        for(int i = size-1; i >= 0;i--)
        {
            System.out.println(num[i] + "     " + step[i]);
        }
    }
// calculate collatz conjecture 
public static long numSteps(long n)
{
    long count = 0; 
    if(n == 1)
    {
      return count; 
    }
    else if ( (n & 1) == 0)
    {
        count = numSteps(n>>1) + 1; 
    }
    else
    {
        count = numSteps((n<<1)+n+1) + 1; 
    }
    return count; 
}

}
